environments/se/tests/src/test/java/org/jboss/weld/environment/se/test/event/options/parallel/ParallelNotificationModeTest.java

```diff
@@ -0,0 +1,100 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jboss.weld.environment.se.test.event.options.parallel;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+import java.util.Set;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.CopyOnWriteArraySet;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+
+import javax.enterprise.event.NotificationOptions;
+
+import org.jboss.arquillian.container.se.api.ClassPath;
+import org.jboss.arquillian.container.test.api.Deployment;
+import org.jboss.arquillian.junit.Arquillian;
+import org.jboss.shrinkwrap.api.Archive;
+import org.jboss.shrinkwrap.api.BeanArchive;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.weld.config.ConfigurationKey;
+import org.jboss.weld.environment.se.Weld;
+import org.jboss.weld.environment.se.WeldContainer;
+import org.jboss.weld.environment.se.test.instance.enhanced.WeldInstanceTest;
+import org.jboss.weld.event.NotificationOptionKeys;
+import org.jboss.weld.event.NotificationOptionKeys.NotificationMode;
+import org.jboss.weld.test.util.Utils;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+/**
+ *
+ * @author Martin Kouba
+ */
+@RunWith(Arquillian.class)
+public class ParallelNotificationModeTest {
+
+    @Deployment
+    public static Archive<?> createTestArchive() {
+        return ClassPath.builder().add(ShrinkWrap.create(BeanArchive.class, Utils.getDeploymentNameAsHash(WeldInstanceTest.class))
+                .addPackage(ParallelNotificationModeTest.class.getPackage())).build();
+    }
+
+    @Test
+    public void testAsyncEventExecutedInDifferentThread() throws InterruptedException {
+        try (WeldContainer container = createWeld()) {
+            BlockingQueue<Message> synchronizer = new LinkedBlockingQueue<>();
+            Set<String> threadNames = new CopyOnWriteArraySet<>();
+            container.event().select(Message.class).fireAsync(() -> threadNames.add(Thread.currentThread().getName()),
+                    NotificationOptions.of(NotificationOptionKeys.MODE, NotificationMode.PARALLEL)).thenAccept(synchronizer::add);
+            Message message = synchronizer.poll(2, TimeUnit.SECONDS);
+            assertNotNull(message);
+            // Eeach observer was notified using a different thread
+            assertEquals(2, threadNames.size());
+        }
+    }
+
+    @Test
+    public void testExceptionPropagated() throws InterruptedException {
+        try (WeldContainer container = createWeld()) {
+            BlockingQueue<Throwable> synchronizer = new LinkedBlockingQueue<>();
+            container.event().select(Message.class).fireAsync(() -> {
+                throw new IllegalStateException(ParallelNotificationModeTest.class.getName());
```

Comment:
What about a test which would throw some exception directly in the observer method?